File: infra-market-server-go/internal/service/activity_service.go
package service

import (
	"encoding/json"
	"log"
	"net/http"

	"github.com/bucketheadv/infra-market/internal/dto"
	"github.com/bucketheadv/infra-market/internal/entity"
	"github.com/bucketheadv/infra-market/internal/repository"
	"github.com/bucketheadv/infra-market/internal/util"
	"gorm.io/gorm"
)

type ActivityService struct {
	db                *gorm.DB
	activityRepo      *repository.ActivityRepository
	templateRepo      *repository.ActivityTemplateRepository
}

func NewActivityService(
	db *gorm.DB,
	activityRepo *repository.ActivityRepository,
	templateRepo *repository.ActivityTemplateRepository,
) *ActivityService {
	return &ActivityService{
		db:                db,
		activityRepo:      activityRepo,
		templateRepo:      templateRepo,
	}
}

// GetActivities 获取活动列表
func (s *ActivityService) GetActivities(query dto.ActivityQueryDto) dto.ApiData[dto.PageResult[dto.ActivityDto]] {
	activities, total, err := s.activityRepo.Page(query)
	if err != nil {
		return dto.Error[dto.PageResult[dto.ActivityDto]]("查询失败", http.StatusInternalServerError)
	}

	// 批量获取模板信息
	templateIDs := make(map[uint64]bool)
	for _, activity := range activities {
		templateIDs[activity.TemplateID] = true
	}

	templateMap := make(map[uint64]*entity.ActivityTemplate)
	for templateID := range templateIDs {
		template, err := s.templateRepo.FindByID(templateID)
		if err == nil {
			templateMap[templateID] = template
		}
	}

	// 转换为DTO
	activityDtos := make([]dto.ActivityDto, len(activities))
	for i, activity := range activities {
		template := templateMap[activity.TemplateID]
		var templateName *string
		if template != nil {
			templateName = &template.Name
		}
		activityDtos[i] = s.convertActivityToDto(&activity, templateName)
	}

	result := dto.PageResult[dto.ActivityDto]{
		Records: activityDtos,
		Total:   total,
		Page:    query.GetPage(),
		Size:    query.GetSize(),
	}

	return dto.Success(result)
}

// GetActivity 获取活动详情
func (s *ActivityService) GetActivity(id uint64) dto.ApiData[dto.ActivityDto] {
	activity, err := s.activityRepo.FindByID(id)
	if err != nil {
		log.Printf("获取活动详情失败，活动ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityDto]("活动不存在", http.StatusNotFound)
	}

	// 获取模板信息
	var templateName *string
	template, err := s.templateRepo.FindByID(activity.TemplateID)
	if err == nil {
		templateName = &template.Name
	}

	activityDto := s.convertActivityToDto(activity, templateName)
	return dto.Success(activityDto)
}

// CreateActivity 创建活动
func (s *ActivityService) CreateActivity(form dto.ActivityFormDto) dto.ApiData[dto.ActivityDto] {
	// 验证模板是否存在
	template, err := s.templateRepo.FindByID(form.TemplateID)
	if err != nil {
		return dto.Error[dto.ActivityDto]("活动模板不存在", http.StatusBadRequest)
	}

	// 验证配置数据
	if err := s.validateConfigData(form.ConfigData, template); err != nil {
		return dto.Error[dto.ActivityDto](err.Error(), http.StatusBadRequest)
	}

	// 序列化配置数据
	configDataJSON, err := s.serializeConfigData(form.ConfigData)
	if err != nil {
		log.Printf("序列化配置数据失败: %v\n", err)
		return dto.Error[dto.ActivityDto]("配置数据格式错误", http.StatusBadRequest)
	}

	status := 1
	if form.Status != nil {
		status = *form.Status
	}

	activity := &entity.Activity{
		Name:        form.Name,
		Description: form.Description,
		TemplateID:  form.TemplateID,
		ConfigData:  configDataJSON,
		Status:      status,
	}

	if err := s.activityRepo.Create(activity); err != nil {
		log.Printf("创建活动失败: %v\n", err)
		return dto.Error[dto.ActivityDto]("创建活动失败", http.StatusInternalServerError)
	}

	activityDto := s.convertActivityToDto(activity, &template.Name)
	return dto.Success(activityDto)
}

// UpdateActivity 更新活动
func (s *ActivityService) UpdateActivity(id uint64, form dto.ActivityFormDto) dto.ApiData[dto.ActivityDto] {
	activity, err := s.activityRepo.FindByID(id)
	if err != nil {
		log.Printf("更新活动失败，活动ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityDto]("活动不存在", http.StatusNotFound)
	}

	// 验证模板是否存在
	if form.TemplateID != 0 {
		template, err := s.templateRepo.FindByID(form.TemplateID)
		if err != nil {
			return dto.Error[dto.ActivityDto]("活动模板不存在", http.StatusBadRequest)
		}

		// 验证配置数据
		if err := s.validateConfigData(form.ConfigData, template); err != nil {
			return dto.Error[dto.ActivityDto](err.Error(), http.StatusBadRequest)
		}
	}

	// 序列化配置数据
	var configDataJSON *string
	if form.ConfigData != nil {
		jsonStr, err := s.serializeConfigData(form.ConfigData)
		if err != nil {
			log.Printf("序列化配置数据失败: %v\n", err)
			return dto.Error[dto.ActivityDto]("配置数据格式错误", http.StatusBadRequest)
		}
		configDataJSON = jsonStr
	}

	activity.Name = form.Name
	if form.Description != nil {
		activity.Description = form.Description
	}
	if form.TemplateID != 0 {
		activity.TemplateID = form.TemplateID
	}
	if configDataJSON != nil {
		activity.ConfigData = configDataJSON
	}
	if form.Status != nil {
		activity.Status = *form.Status
	}

	if err := s.activityRepo.Update(activity); err != nil {
		log.Printf("更新活动失败，活动ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityDto]("更新活动失败", http.StatusInternalServerError)
	}

	// 获取模板名称
	var templateName *string
	template, err := s.templateRepo.FindByID(activity.TemplateID)
	if err == nil {
		templateName = &template.Name
	}

	activityDto := s.convertActivityToDto(activity, templateName)
	return dto.Success(activityDto)
}

// DeleteActivity 删除活动
func (s *ActivityService) DeleteActivity(id uint64) dto.ApiData[any] {
	_, err := s.activityRepo.FindByID(id)
	if err != nil {
		log.Printf("删除活动失败，活动ID: %d, 错误: %v\n", id, err)
		return dto.Error[any]("活动不存在", http.StatusNotFound)
	}

	if err := s.activityRepo.Delete(id); err != nil {
		log.Printf("删除活动失败，活动ID: %d, 错误: %v\n", id, err)
		return dto.Error[any]("删除活动失败", http.StatusInternalServerError)
	}

	return dto.Success[any](nil)
}

// UpdateActivityStatus 更新活动状态
func (s *ActivityService) UpdateActivityStatus(id uint64, status int) dto.ApiData[dto.ActivityDto] {
	activity, err := s.activityRepo.FindByID(id)
	if err != nil {
		log.Printf("更新活动状态失败，活动ID: %d, 错误: %v\n", id, err)
		return dto.Error[dto.ActivityDto]("活动不存在", http.StatusNotFound)
	}

	activity.Status = status
	if err := s.activityRepo.Update(activity); err != nil {
		log.Printf("更新活动状态失败，活动ID: %d, 状态: %d, 错误: %v\n", id, status, err)
		return dto.Error[dto.ActivityDto]("更新状态失败", http.StatusInternalServerError)
	}

	// 获取模板名称
	var templateName *string
	template, err := s.templateRepo.FindByID(activity.TemplateID)
	if err == nil {
		templateName = &template.Name
	}

	activityDto := s.convertActivityToDto(activity, templateName)
	return dto.Success(activityDto)
}

// convertActivityToDto 转换活动实体为DTO
func (s *ActivityService) convertActivityToDto(activity *entity.Activity, templateName *string) dto.ActivityDto {
	configData := s.parseConfigData(activity.ConfigData)

	return dto.ActivityDto{
		ID:           activity.ID,
		Name:         activity.Name,
		Description:  activity.Description,
		TemplateID:   activity.TemplateID,
		TemplateName: templateName,
		ConfigData:   configData,
		Status:       activity.Status,
		CreateTime:   util.Format(&activity.CreateTime),
		UpdateTime:   util.Format(&activity.UpdateTime),
	}
}

// serializeConfigData 序列化配置数据为JSON字符串
func (s *ActivityService) serializeConfigData(configData map[string]any) (*string, error) {
	if configData == nil {
		return nil, nil
	}

	jsonBytes, err := json.Marshal(configData)
	if err != nil {
		return nil, err
	}

	jsonStr := string(jsonBytes)
	return &jsonStr, nil
}

// parseConfigData 解析配置数据JSON字符串
func (s *ActivityService) parseConfigData(configDataJSON *string) map[string]any {
	if configDataJSON == nil || *configDataJSON == "" {
		return nil
	}

	var configData map[string]any
	if err := json.Unmarshal([]byte(*configDataJSON), &configData); err != nil {
		log.Printf("解析配置数据失败: %v\n", err)
		return nil
	}

	return configData
}

// validateConfigData 验证配置数据是否符合模板的字段配置
func (s *ActivityService) validateConfigData(configData map[string]any, template *entity.ActivityTemplate) error {
	if template.Fields == nil || *template.Fields == "" {
		return nil
	}

	var fields []dto.ActivityTemplateFieldDto
	if err := json.Unmarshal([]byte(*template.Fields), &fields); err != nil {
		log.Printf("解析模板字段配置失败: %v\n", err)
		return nil
	}

	// 验证必填字段
	for _, field := range fields {
		if field.Required != nil && *field.Required {
			fieldName := field.Name
			if fieldName == "" {
				continue
			}

			value, exists := configData[fieldName]
			if !exists || value == nil {
				displayName := fieldName
				if field.Label != nil && *field.Label != "" {
					displayName = *field.Label + "（" + fieldName + "）"
				}
				return &ValidationError{Message: "字段 " + displayName + " 为必填项，不能为空"}
			}

			// 检查字符串是否为空
			if strValue, ok := value.(string); ok && strValue == "" {
				displayName := fieldName
				if field.Label != nil && *field.Label != "" {
					displayName = *field.Label + "（" + fieldName + "）"
				}
				return &ValidationError{Message: "字段 " + displayName + " 为必填项，不能为空"}
			}

			// 检查数组是否为空
			if arrValue, ok := value.([]any); ok && len(arrValue) == 0 {
				displayName := fieldName
				if field.Label != nil && *field.Label != "" {
					displayName = *field.Label + "（" + fieldName + "）"
				}
				return &ValidationError{Message: "字段 " + displayName + " 为必填项，不能为空"}
			}
		}
	}

	return nil
}

// ValidationError 验证错误
type ValidationError struct {
	Message string
}

func (e *ValidationError) Error() string {
	return e.Message
}

File: infra-market-server-vertx/src/main/kotlin/io/infra/market/vertx/service/ActivityService.kt
package io.infra.market.vertx.service

import com.google.inject.Inject
import com.google.inject.Singleton
import io.infra.market.vertx.config.JacksonConfig
import io.infra.market.vertx.dto.ActivityDto
import io.infra.market.vertx.dto.ActivityFormDto
import io.infra.market.vertx.dto.ActivityQueryDto
import io.infra.market.vertx.dto.ApiData
import io.infra.market.vertx.dto.PageResultDto
import io.infra.market.vertx.entity.Activity
import io.infra.market.vertx.extensions.awaitForResult
import io.infra.market.vertx.repository.ActivityDao
import io.infra.market.vertx.util.TimeUtil
import io.vertx.sqlclient.Pool
import io.vertx.sqlclient.Tuple
import org.slf4j.LoggerFactory

/**
 * 活动服务
 * 
 * 规则1：任何调用 xxx.awaitForResult() 的函数，必须用 suspend 修饰
 */
@Singleton
class ActivityService @Inject constructor(
    private val activityDao: ActivityDao,
    private val pool: Pool
) {
    
    private val logger = LoggerFactory.getLogger(ActivityService::class.java)
    
    suspend fun getActivities(query: ActivityQueryDto): ApiData<PageResultDto<ActivityDto>> {
        val page = query.page ?: 1
        val size = query.size ?: 10
        
        val pageResult = activityDao.findPage(
            name = query.name,
            templateId = query.templateId,
            status = query.status,
            page = page,
            size = size
        )
        
        // 批量获取模板名称
        val templateIds = pageResult.records.mapNotNull { it.templateId }.distinct()
        val templateNameMap = mutableMapOf<Long, String?>()
        templateIds.forEach { templateId ->
            val templateName = getTemplateName(templateId)
            templateNameMap[templateId] = templateName
        }
        
        val activityDtos = pageResult.records.map { activity ->
            convertToDto(activity, templateNameMap[activity.templateId])
        }
        
        val result = PageResultDto(
            records = activityDtos,
            total = pageResult.total,
            page = pageResult.page,
            size = pageResult.size
        )
        
        return ApiData.success(result)
    }
    
    suspend fun getActivity(id: Long): ApiData<ActivityDto> {
        val activity = activityDao.findById(id) ?: return ApiData.error("活动不存在", 404)
        
        val templateName = activity.templateId?.let { getTemplateName(it) }
        val activityDto = convertToDto(activity, templateName)
        
        return ApiData.success(activityDto)
    }
    
    suspend fun createActivity(form: ActivityFormDto): ApiData<ActivityDto> {
        // 验证模板是否存在
        val templateId = form.templateId
        if (templateId == null) {
            return ApiData.error("模板ID不能为空", 400)
        }
        
        val templateExists = checkTemplateExists(templateId)
        if (!templateExists) {
            return ApiData.error("活动模板不存在", 400)
        }
        
        val activity = convertToEntity(form)
        activity.status = form.status ?: 1
        
        val activityId = activityDao.save(activity)
        activity.id = activityId
        
        val templateName = getTemplateName(templateId)
        val activityDto = convertToDto(activity, templateName)
        
        return ApiData.success(activityDto)
    }
    
    suspend fun updateActivity(id: Long, form: ActivityFormDto): ApiData<ActivityDto> {
        val existingActivity = activityDao.findById(id) ?: return ApiData.error("活动不存在", 404)
        
        // 验证模板是否存在
        val formTemplateId = form.templateId
        if (formTemplateId != null) {
            val templateExists = checkTemplateExists(formTemplateId)
            if (!templateExists) {
                return ApiData.error("活动模板不存在", 400)
            }
        }
        
        existingActivity.name = form.name
        existingActivity.description = form.description
        if (formTemplateId != null) {
            existingActivity.templateId = formTemplateId
        }
        existingActivity.configData = serializeConfigData(form.configData)
        if (form.status != null) {
            existingActivity.status = form.status
        }
        
        activityDao.updateById(existingActivity)
        
        val templateId = existingActivity.templateId
        val templateName = templateId?.let { getTemplateName(it) }
        val activityDto = convertToDto(existingActivity, templateName)
        
        return ApiData.success(activityDto)
    }
    
    suspend fun deleteActivity(id: Long): ApiData<Unit> {
        val activity = activityDao.findById(id) ?: return ApiData.error("活动不存在", 404)
        
        activityDao.deleteById(id)
        
        return ApiData.success()
    }
    
    suspend fun updateActivityStatus(id: Long, status: Int): ApiData<ActivityDto> {
        val activity = activityDao.findById(id) ?: return ApiData.error("活动不存在", 404)
        
        activity.status = status
        activityDao.updateById(activity)
        
        val templateName = activity.templateId?.let { getTemplateName(it) }
        val activityDto = convertToDto(activity, templateName)
        
        return ApiData.success(activityDto)
    }
    
    /**
     * 获取模板名称
     */
    private suspend fun getTemplateName(templateId: Long): String? {
        return try {
            val rows = pool.preparedQuery("SELECT name FROM activity_template WHERE id = ?")
                .execute(Tuple.of(templateId))
                .awaitForResult()
            if (rows.size() > 0) {
                rows.first().getString("name")
            } else {
                null
            }
        } catch (e: Exception) {
            logger.warn("查询模板名称失败，模板ID: $templateId", e)
            null
        }
    }
    
    /**
     * 检查模板是否存在
     */
    private suspend fun checkTemplateExists(templateId: Long): Boolean {
        return try {
            val rows = pool.preparedQuery("SELECT COUNT(*) as count FROM activity_template WHERE id = ?")
                .execute(Tuple.of(templateId))
                .awaitForResult()
            rows.first().getLong("count") > 0
        } catch (e: Exception) {
            logger.warn("检查模板是否存在失败，模板ID: $templateId", e)
            false
        }
    }
    
    /**
     * 转换为DTO
     */
    private fun convertToDto(activity: Activity, templateName: String?): ActivityDto {
        val configData = parseConfigData(activity.configData)
        
        return ActivityDto(
            id = activity.id,
            name = activity.name,
            description = activity.description,
            templateId = activity.templateId,
            templateName = templateName,
            configData = configData,
            status = activity.status,
            createTime = TimeUtil.format(activity.createTime),
            updateTime = TimeUtil.format(activity.updateTime)
        )
    }
    
    /**
     * 转换为实体
     */
    private fun convertToEntity(form: ActivityFormDto): Activity {
        val configDataJson = serializeConfigData(form.configData)
        
        return Activity(
            name = form.name,
            description = form.description,
            templateId = form.templateId,
            configData = configDataJson,
            status = form.status
        )
    }
    
    /**
     * 序列化配置数据为JSON字符串
     */
    private fun serializeConfigData(configData: Map<String, Any>?): String? {
        return try {
            if (configData.isNullOrEmpty()) {
                null
            } else {
                JacksonConfig.objectMapper.writeValueAsString(configData)
            }
        } catch (e: Exception) {
            logger.error("序列化配置数据失败: ${e.message}", e)
            null
        }
    }
    
    /**
     * 解析配置数据JSON字符串
     */
    private fun parseConfigData(configDataJson: String?): Map<String, Any>? {
        return try {
            if (configDataJson.isNullOrBlank()) {
                null
            } else {
                @Suppress("UNCHECKED_CAST")
                JacksonConfig.objectMapper.readValue(configDataJson, Map::class.java) as? Map<String, Any>
            }
        } catch (e: Exception) {
            logger.error("解析配置数据失败: ${e.message}", e)
            null
        }
    }
}
